README.md
# UnifiedPush Server Admin Client

![Build](https://github.com/aerogear/unifiedpush-admin-client/workflows/build/badge.svg)
[![Coverage Status](https://coveralls.io/repos/github/aerogear/unifiedpush-admin-client/badge.svg?branch=master)](https://coveralls.io/github/aerogear/unifiedpush-admin-client)
[![Codacy Badge](https://app.codacy.com/project/badge/Grade/af0c8d43f45a418a8844b68969b7efd2)](https://www.codacy.com/gh/aerogear/unifiedpush-admin-client?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=aerogear/unifiedpush-admin-client&amp;utm_campaign=Badge_Grade)

The _UnifiedPush Server Admin_ library allows to admin the UPS by javascript or typescript code.

Supported features are:

**Applications**
* **create**: allows to create an application initialised with all the provided properties
* **delete**: allows to delete all the applications matching given criteria
* **update**: allows to update a given application
* **search**: allows to search for all the applications matching given criteria

**Variants**
* **create**: allows to create a variant initialised with all the specified properties
* **delete**: allows to delete all the variants matching given criteria
* **update**: allows to update a given variant
* **search**: allows to search for all the variants matching given criteria
* **renew secret**: allows to renew the secret of a given variant

## Getting started
### Cloning the repository

```bash
git clone https://github.com/aerogear/unifiedpush-admin-client.git
cd unifiedpush-admin-client
```

#### Compiling the code

```bash
npm run compile
```

#### Running the tests

```bash
npm run test
```

### Using the library

The entry point of the library is the _UpsAdminClient_ class. Its constructor takes two arguments:
1. **serverURL**: The URL of the _UnifiedPushServer_. The url must be complete of protocol and port.  
   Example:  
   `http://localhost:9999` 
2. **credentials**: This parameter is be used to authenticate before trying to call the UPS REST endpoints. Two types of credentials
can be specified:
   * Basic: this is just username/password used for basic authentication
   * Keycloak: this is composed of **username/password**, **realm**, **client_id** and **keycloak URL**. Optionally, if you already have a valid bearer token, it
   can be specified here, so that the library won't try to authenticate again.
   
**Example without credentials:** 
```typescript
const upsadm = new UpsAdminClient('http://localhosrt:9999');
```

**Example with basic authentication:**  
```typescript
const upsadm = new UpsAdminClient('http://localhosrt:9999', 
        {username: 'username', password: 'password', type: 'basic'});
```

**Example with keycloak authentication:**
```typescript
const upsadm = new UpsAdminClient('http://localhosrt:9999', 
        {
          kcUrl: 'http://172.18.0.2:8080',
          username: 'username',
          password: 'password',
          type: 'keycloak',
        });
```

You will notice that we didn't specify values for `client_id` and `realm`: they will default to `unified-push-server-js` 
and `aerogear` accordingly.

#### Managing applications

All the methods to be used to manage applications can be found inside the `applications` namespace.
To get the list of the applications, we will use
```typescript
const searchResult = await upsadm.applications.search().execute();
console.log(`Total number of applications: ${searchResult.total}`);
console.log('Applications:', searchResult.list)
```

The search provides a _fluid interface_ that allows to filter the results.
For example, to search all the applications named 'TEST' we can do:

```typescript
const searchResult = await upsadm.applications.search()
    .withName('TEST')
    .execute();
```

More than one filter can be specified.

**WARNING** The result is returned one page at a time. Each page will contain 10 applications. To specify which page to return, the 
`page` parameter must be specified:

```typescript
const searchResult = await upsadm.applications.search()
    .withName('TEST')
    .page(3)
    .execute();
```
The returned value will be a dictionary containing the following keys:
* **total**: the total number of apps that match the given filter
* **list**: the current result page

To create an application we will use the `create` method:
```typescript
const app = await upsadm.applications.create('newApp').execute();
```
The returned `app` object will have all the `id` and `pushApplicationId` fields populated by the server.

To update an application we will use the `update` method:

```typescript
const apps = await upsadm.applications.update(appId)
    .withName('newName')
    .execute();
```  

#### Managing variants

All the methods to be used to manage variants can be found inside the `variants` namespace.
To get the list of the variants, we will use
```typescript
const apps = await upsadm.variants.search(appId).execute();
```
where `appId` is the id of the application owning the variants.

The search command provides a fluent interface that can be used to add filters to the search:
```typescript
const apps = await upsadm.variants.search(appId)
    .withName('my variant')
    .execute();
```
**NOTE** You are not limited to use only one filter: if you specify more than one, they will be all applied. The only 
exception is the `id` filter: if specified, all the other filters will be ignored.

To create a variant we will use the `create` method:
```typescript
const variant = upsadm.variants.android.
    create(appId)
    .withName('MyNewVariant')
    .withGoogleKey('googleKey')
    .withProjectNumber('12345')
    .execute();
```
The returned `variant` object will have all the `pushApplicationID` field populated by the server.

To delete variants, we will use the `delete` method:

```typescript
const apps = upsadm.variants.delete(appId).execute();
```

If you don't want to delete all the variants, you can decide what to delete by using the fluent interface:

```typescript
const apps = upsadm.variants.delete(appId)
    .withName('apptodelete')
    .execute();
```

If you want to work on variants of a specified type, you can use the scoped version of the delete method:

```typescript
const apps = upsadm.variants.android.delete(appId)
    .withName('apptodelete')
    .execute();
```


CHANGELOG.md
## [4.5.3](https://github.com/aerogear/unifiedpush-admin-client/compare/4.5.2...4.5.3) (2022-07-28)


### Bug Fixes

* **deps:** update dependency axios to v0.27.2 ([bed056a](https://github.com/aerogear/unifiedpush-admin-client/commit/bed056a86e5c52629e9cead233c10ad8aca7df7d))



## [4.5.2](https://github.com/aerogear/unifiedpush-admin-client/compare/4.5.1...4.5.2) (2021-09-24)


### Bug Fixes

* **deps:** update dependency axios to v0.21.4 ([00a5c9d](https://github.com/aerogear/unifiedpush-admin-client/commit/00a5c9da7b2147eeebd1b1d3d86a640ac3e56bcc))
* package.json & package-lock.json to reduce vulnerabilities ([c09a4dc](https://github.com/aerogear/unifiedpush-admin-client/commit/c09a4dc651930e6c75cba2f5113d4d3c886b1297))



## [4.5.1](https://github.com/aerogear/unifiedpush-admin-client/compare/4.5.0...4.5.1) (2021-03-04)


### Bug Fixes

* **deps:** update dependency axios to v0.21.1 ([fb4a36e](https://github.com/aerogear/unifiedpush-admin-client/commit/fb4a36e4dc6ae423104180bac622bb4e175e904a))
* **deps:** update dependency form-data to v4 ([d7fa95a](https://github.com/aerogear/unifiedpush-admin-client/commit/d7fa95ad2cf23c0f92204136a0de630269d1e052))



# [4.5.0](https://github.com/aerogear/unifiedpush-admin-client/compare/4.4.3...4.5.0) (2020-12-15)


### Features

* 🎸 now when searching apps you can get the complete list ([53ad8ac](https://github.com/aerogear/unifiedpush-admin-client/commit/53ad8ac915e9db1d7717fa39c6fc68d0c8147b9d))



## [4.4.3](https://github.com/aerogear/unifiedpush-admin-client/compare/4.4.2...4.4.3) (2020-12-14)


### Bug Fixes

* 🐛 return value of RenewVapidKeysCommand.exec was wrong ([83a8cb0](https://github.com/aerogear/unifiedpush-admin-client/commit/83a8cb002edb86836b8cf89aa48d7c117830f4da))
* 🐛 the LoadMetricsCommand was returning strings ([608e399](https://github.com/aerogear/unifiedpush-admin-client/commit/608e399df30643763555e8604aa81f7776c5d068))





src/main.ts
import { UnifiedPushAdminClient } from './UnifiedPushAdminClient';

async function run() {
  // const app = await new UPSClient().applications.create(newApp);
  // console.log('NewApp: ', app);

  // const apps = await new UnifiedPushClient().applications.find({
  //   name: 'New Ziccardi',
  // });
  // console.log('apps', apps);

  //B868CC08-BCC8-4A0A-B21E-1AC56AF0C734
  // const variants = await new UnifiedPushClient('http://localhost:9999', {
  //   kcUrl: 'http://172.18.0.2:8080',
  //   username: 'admin',
  //   password: '123',
  // }).variants.find('58939fc1-8868-46f5-95e3-0637ea97f9f4');

  const variants = await new UnifiedPushAdminClient('http://localhost:9999', {
    kcUrl: 'http://172.18.0.2:8080',
    username: 'admin',
    password: '123',
    type: 'keycloak',
  }).variants.find('58939fc1-8868-46f5-95e3-0637ea97f9f4');
  console.log('variants', variants);
}

run().catch(err => console.log('err', err));
